src/main/java/org/kiwiproject/registry/consul/server/ConsulRegistryService.java

```diff
@@ -0,0 +1,220 @@
+     */
+    private static final long UNREGISTER_RETRY_DELAY = 3;
+
+    /**
+     * Time unit for delay when un-registering if a call to Eureka fails.
+     */
+    private static final TimeUnit UNREGISTER_RETRY_DELAY_UNIT = TimeUnit.SECONDS;
+
+    /**
+     * Maximum number of attempts we will make to un-register from Eureka (via DELETE request).
+     */
+    private static final int MAX_UNREGISTER_ATTEMPTS = 5;
+
+    private final Consul consul;
+    private final ConsulRegistrationConfig config;
+    private final SimpleRetryer registerRetryer;
+    private final SimpleRetryer unregisterRetryer;
+
+    @VisibleForTesting
+    final AtomicReference<ServiceInstance> registeredService;
+
+    public ConsulRegistryService(Consul consul, ConsulRegistrationConfig config, KiwiEnvironment environment) {
+        this.consul = consul;
+        this.config = config;
+        this.registeredService = new AtomicReference<>();
+
+        this.registerRetryer = SimpleRetryer.builder()
+                .environment(environment)
+                .maxAttempts(MAX_REGISTRATION_ATTEMPTS)
+                .retryDelayTime(RETRY_DELAY)
+                .retryDelayUnit(RETRY_DELAY_UNIT)
+                .build();
+
+        this.unregisterRetryer = SimpleRetryer.builder()
+                .environment(environment)
+                .maxAttempts(MAX_UNREGISTER_ATTEMPTS)
+                .retryDelayTime(UNREGISTER_RETRY_DELAY)
+                .retryDelayUnit(UNREGISTER_RETRY_DELAY_UNIT)
+                .build();
+    }
+
+    @Override
+    public ServiceInstance createCandidateFrom(ServiceInfo serviceInfo) {
+        return ServiceInstance.fromServiceInfo(serviceInfo)
+                .withStatus(ServiceInstance.Status.UP);
+    }
+
+    @Override
+    public ServiceInstance register(ServiceInstance serviceToRegister) {
+        checkState(!isRegistered(), "Cannot register. Already managing a registered instance");
+
+        var registration = fromServiceInstance(serviceToRegister);
+
+        var service = registerRetryer.tryGetObject(() -> {
+            consul.agentClient().register(registration);
+            return serviceToRegister.withInstanceId(registration.getId());
+        });
+
+        if (service.isEmpty()) {
+            LOG.error("Registration failed for service {} with id {}. See logs above", serviceToRegister.getServiceName(), registration.getId());
+            var errMsg = format("Unable to register service %s, id %s with Consul after %s attempts",
+                    serviceToRegister.getServiceName(), registration.getId(), MAX_REGISTRATION_ATTEMPTS);
+            throw new RegistrationException(errMsg);
+        }
+
+        registeredService.set(service.get());
+        return registeredService.get().toBuilder().build();
+    }
+
+    @Override
+    public ServiceInstance updateStatus(ServiceInstance.Status newStatus) {
+        LOG.info("Ignoring status update as Consul handles it's own status through it's health check process");
```

Comment:
Leave this at INFO...or change to WARN? Are you expecting that clients should _not_ call this if using Consul?

Reply:
Changing this to WARN based on adding `checkState` and in case we ever need to actually do something.  Users should not be calling this